libcdb: Load debug info and unstrip libc binary

peace-maker opened the pull request:
The debug info is obtained from a debuginfod server maintained by systemtap.org. elfutils is used to unstrip the binary and merge the debug info back.

Try to unstrip every libc binary downloaded by the libcdb module by default and add a new `unstrip_libc(filename)` function to handle other libc binaries.

Closes #1818

peace-maker committed:
libcdb: Load debug info and unstrip libc binary

The debug info is obtained from a debuginfod server maintained by systemtap.org. elfutils is used to unstrip the binary and merge the debug info back.

Try to unstrip every libc binary downloaded by the libcdb module by default. #1818

peace-maker committed:
Install elfutils on CI

peace-maker committed:
Update CHANGELOG.md

peace-maker commented:
Hmm, I'm not sure what GHA is doing to the filesystem. Sometimes it appears to be fine e.g. [here](https://github.com/peace-maker/pwntools/runs/1997583526?check_suite_focus=true#step:16:444).
Those are tests against an external service again, so maybe just not run them?

peace-maker commented:
It appears to be a problem with the elfutils version installed. The version installed on focal `0.176-1.1build1` failed on the last run. The bionic version `0.170-0.4ubuntu0.1` is fine and my local version `0.182` as well.

heapcrash commented on pwnlib/libcdb.py:
```diff
@@ -124,16 +112,122 @@ def search_by_hash(hex_encoded_id, hash_type='build_id'):
+    log.info_once("Using cached data from %r", cache)
+    return cache, True
+
+def unstrip_libc(filename):
```
This needs doctests for a few examples

peace-maker replied:
"A few examples" = unstripping more than just one libc?
I've added a few more tests, is this what you had in mind?

heapcrash commented on pwnlib/libcdb.py:
```diff
@@ -124,16 +112,122 @@ def search_by_hash(hex_encoded_id, hash_type='build_id'):
     return cache
 
+def _check_elf_cache(cache_type, hex_encoded_id, hash_type):
+    # Ensure that the cache directory exists
```
This needs doctests, which should:

* Show that there is no cached file for a given hash
* Fetch that specific hash
* Show that now there IS a cached file for the hash

heapcrash reviewed (changes requested):
Looks like CI is failing, and I've got a few suggestions for doctests

heapcrash commented:
Any chance for updates to address the requested changes?

peace-maker committed:
Add more tests

Arusekk committed:
Merge branch 'dev' into libcdb_unstrip

peace-maker committed:
Fix negative test

peace-maker commented:
Since the tests seem to fail on certain buggy elfutils versions installed on the CI workers, I'm not sure how to persue with this.

Arusekk committed:
Merge branch 'dev' into libcdb_unstrip

Arusekk committed:
Merge branch 'dev' into libcdb_unstrip

Arusekk commented:
please try to print the full error message in case of failure, so that we can debug the buggy CI

peace-maker committed:
Raise an exception when eu-unstrip fails

This will hopefully get logged during doctests.

Arusekk commented:
So it says invalid section alignment... whatever that means..

peace-maker commented:
:( should we build elfutils from source in CI to work around this? It appears to be fixed in later versions like the one in 20.10.

heapcrash commented:
> :( should we build elfutils from source in CI to work around this? It appears to be fixed in later versions like the one in 20.10.

I don't think we want to include functionality that requires bespoke builds / anything non-standard on 18.04 / 20.04

@Arusekk?

peace-maker commented:
I think only the version packaged on 20.04 is buggy with the libc we fetch in the test. 18.04 and 20.10+ is fine. I'll try to switch to a different libc binary that works on 20.04 too for the doctests. It'd be a shame to strip this feature just because a buggy package version is being used atm.

heapcrash commented:
Agreed, it's a great feature to have.  If it mostly-works and works in most
situations (and we can add tests) we can just deal with regressions
manually.

Pwntools used to distribute .dpkg for binutils for alternate architectures
[1] because the standard package managers didn't have e.g. mip64
cross-assembler.

[1]: https://launchpad.net/~pwntools/+archive/ubuntu/binutils


On Tue, Jun 8, 2021 at 2:51 PM peace-maker ***@***.***> wrote:

> I think only the version packaged on 20.04 is buggy with the libc we fetch
> in the test. 18.04 and 20.10+ is fine. I'll try to switch to a different
> libc binary that works on 20.04 too for the doctests. It'd be a shame to
> strip this feature just because a buggy package version is being used atm.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Gallopsled/pwntools/pull/1828#issuecomment-857066247>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APYTIFP5CK344H7OW4Z2BXDTRZYDHANCNFSM4YK3XWXA>
> .
>

peace-maker committed:
Install elfutils 0.181 from groovy

Use APT Pinning to select the newer elfutils version from groovy instead of the buggy 0.176 version available in focal.

peace-maker committed:
Merge branch 'dev' into libcdb_unstrip

peace-maker committed:
Fix missing import of requests

pylint?

peace-maker commented:
I've fixed the CI by installing a newer elfutils package version from Ubuntu groovy explicitly now while the worker is still on focal.

Arusekk committed:
Merge branch 'dev' into libcdb_unstrip

Arusekk commented:
The CI introduced by this now breaks. I don't know why yet.

peace-maker commented:
Since ubuntu groovy 20.10 is EOL the repository was removed and we can't fetch the elfutils package from a newer ubuntu version anymore. The runners use 20.04 which uses an older libc than the newer ubuntu versions 21.04 and 21.10, so we can't use the same trick to just grab the newer elfutils package from those distributions.

I guess we'll have to resort to building elfutils and eu-unstrip ourselves or disable those tests until the runners are updated to some newer base image.
I started working on building elfutils in CI, but got stuck and didn't had time to investigate further yet. Building the newer version manually in CI causes the tests to fail in the same way as if we'd be using the old binary in the official package. It might be due to the binary still using the old libelf libary instead of the newer one.
